### PR TITLE
add experimental lazy compilation

### PR DESCRIPTION
**Pull request by sokra**
This adds an experimental feature which allows to compile only the part of the application which is currently in use. Additional `import()`s which are currently not in use do not affect compilation performance when this feature is enabled.

When used webpack opens an additional communication channel between client and compiler via Server-Send-Events. With this channel the currently used `import()`s are told the compiler. The compiler will include these `import()`s in the next compilation. When in watch mode and new `import()`s are communicated to the compiler, it will automatically recompile.

When Hot Module Replacement is not enabled, manual refresh of the page is required.

This also works when targeting node.js. Including HMR in node.js.

After 2 minutes not using a `import()` it be garbage collected. Once compiled modules will stay in cache, so compiling it again is probably pretty fast. It's recommended to combine this with the Persistent Cache to keep the cache even after restarting.

**What kind of change does this PR introduce?**
feature

**Did you add tests for your changes?**
yes, but that could be improved.

**Does this PR introduce a breaking change?**
no

**What needs to be documented once your changes are merged?**
* add `experiments.lazyCompilation`: Compiles `import()` only when requested using HMR.

Prior art
* next.js detects the used routes and adds entrypoint on demand, which makes unused routes not influencing the compilation
* https://github.com/liximomo/lazy-compile-webpack-plugin
  * uses timestamp write to disk to invalidate build
  * uses a loader instead of a separate Module

cc @liximomo

### CODE REVIEW

File: lib/hmr/lazyCompilationBackend.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+ * @param {Compiler} compiler compiler
+ * @param {string} client client reference
+ * @param {function(Error?, any?): void} callback callback
+ * @returns {void}
+ */
+module.exports = (compiler, client, callback) => {
+	const logger = compiler.getInfrastructureLogger("LazyCompilationBackend");
+	const activeModules = new Map();
+	const prefix = "/lazy-compilation-using-";
+	const server = http.createServer((req, res) => {
```
[the remark this answers is not in the document]
Comment (sokra) [reply lib/hmr/lazyCompilationBackend.js:22]: That would require providing a cert and key. We can definitely add that before this leaves experimental status.

**Comment by liximomo** [comment]
I am so happy that webpack will integrate this feature into its core. 🎉 🎉 🎉

**Comment by sanriqing** [comment]
it's so cooooool!

**Comment by sokra** [comment]
Great that you like it. Could you provide more feedback/details?

**Comment by sanriqing** [comment]
> Great that you like it. Could you provide more feedback/details?

Sure! i will test it in weekend and give some feedback.

**Comment by pcostanz** [comment]
Enabling this feature seems to cause our (very large) build to take longer, reload times on a small react component seem unchanged. The app has many (hundreds) of dynamic import chunks. Let me know if I can provide any other info.

Relevant dependencies:

```
    "@pmmmwh/react-refresh-webpack-plugin": "^0.4.3",
    "express": "^4.17.1",
    "react-refresh": "^0.8.3",
    "webpack": "^5.21.2",
    "webpack-bundle-analyzer": "^4.3.0",
    "webpack-dev-middleware": "^4.1.0",
    "webpack-hot-middleware": "^2.25.0",

```


Before
```
webpack 5.21.2 compiled with 1 warning in 128680 ms
webpack building...
webpack built 64dcfd39f0279a778a78 in 6257ms
```

After
```
webpack 5.21.2 compiled with 1 warning in 202855 ms
webpack building...
webpack built 9fb56e85ce36e9f5b063 in 6227ms
```

**Comment by alexander-akait** [comment]
`webpack-hot-middleware` doesn't work with this

**Comment by sokra** [comment]
Not sure. Please open an issue with a little bit more info.

**Comment by ramasilveyra** [comment]
When lazy compilation is enabled with next.js you need to request the page 3 times in order to be rendered properly.

Bug repro: https://github.com/ramasilveyra/bug-repro-next-lazy-compilation

Where should I report this issue? here in webpack or in next.js?

**Comment by alexander-akait** [comment]
@ramasilveyra Looks like bug in next.js

File: lib/hmr/lazyCompilationBackend.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+		});
+		req.socket.setNoDelay(true);
+		res.writeHead(200, {
+			"content-type": "text/event-stream",
+			"Access-Control-Allow-Origin": "*"
+		});
+		res.write("\n");
+		let moduleActivated = false;
+		for (const key of keys) {
+			const oldValue = activeModules.get(key) || 0;
```
Comment (ramasilveyra) [inline lib/hmr/lazyCompilationBackend.js:45]: Why server sent events are used for this?

**Comment by ramasilveyra** [comment]
This feature is amazing!

One page of one private big app using next.js + webpack 5 was taking ~3:30min to compile without cache in development.

With this change first build without cache of this page takes ~12s! 🚀

**Comment by ramasilveyra** [comment]
The only problem that I found is that sometimes in some engineers laptops one sporadic issue appears. I couldn't repro in a reliable way yet.

It seems that it's happening when engineers open pages multiple times, creating multiple SSE.

This is the error in console:

Screen Shot 2021-03-08 at 8 33 27 PM

But by reading the code from the backend and client lazy compilation communication , it seems that SSE aren't needed.

Do you know why SSE are used?

I've created a draft PR replacing SSE with a regulat HTTP GET request here: https://github.com/webpack/webpack/pull/12838

**Comment by sokra** [comment]
SSE is used because it's an easy way to create a persistent connection. A single ping request is not enough, you would need to sent a ping every minute or so. The client need to keep the keep the modules in the compilation as long the page is open. SSE is designed for this kind of persistent connection. Pinging via interval might get throttled by the browser when the page inactive. But I see the potential problem with too many open connections so using ping as fallback seems reasonable. Or using multiple ports on the server.

**Comment by zhangchao828** [comment]
webpack-hot-middleware doesn't work with this?

**Comment by await-ovo** [comment]
should turn  lazy compilation off by default in production mode?

**Comment by JounQin** [comment]
image

It seems blocked by CORS.

**Comment by alexander-akait** [comment]
@JounQin hm, weird, I think we should add ability to set headers, can you provide small example?

**Pull request merged**